src/DomainLib.Projections.EventStore/EventStoreDroppedSubscriptionHandler.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Threading.Tasks;
+using DomainLib.Common;
+using EventStore.ClientAPI;
+using Microsoft.Extensions.Logging;
+
+namespace DomainLib.Projections.EventStore
+{
+    public class EventStoreDroppedSubscriptionHandler
+    {
+        private readonly Action _stop;
+        private readonly Func<Task> _resubscribe;
+        private static readonly ILogger<EventStoreDroppedSubscriptionHandler> Log = Logger.CreateFor<EventStoreDroppedSubscriptionHandler>();
+
+        private int _maxSubscribeAttempts = 3;
+        private int _subscribeAttempts;
+        private TimeSpan _backOffTimeSpan = TimeSpan.FromSeconds(1);
+
+        public EventStoreDroppedSubscriptionHandler(Action stop, Func<Task> resubscribe)
+        {
+            _stop = stop;
+            _resubscribe = resubscribe;
+        }
+        
+        public void HandleDroppedSubscription(SubscriptionDropReason reason, Exception exception)
+        {
+            switch (reason)
+            {
+                case SubscriptionDropReason.UserInitiated:
+                    Log.LogInformation("User initiated EventStore subscription drop. Stopping event publisher");
+                    _stop();
+                    break;
+                case SubscriptionDropReason.NotAuthenticated:
+                case SubscriptionDropReason.AccessDenied:
+                    Log.LogCritical(exception,
+                                    $"Error subscribing to EventStore. Stopping event publisher. Reason {reason}");
+                    _stop();
+                    break;
+                case SubscriptionDropReason.SubscribingError:
+                case SubscriptionDropReason.CatchUpError:
+                case SubscriptionDropReason.ServerError:
+                case SubscriptionDropReason.EventHandlerException:
+                case SubscriptionDropReason.Unknown:
+                case SubscriptionDropReason.NotFound:
+                    Log.LogCritical(exception,
+                                    $"Exception occurred in EventStore subscription. Stopping event publisher. Reason {reason}");
+                    _stop();
+                    break;
+                case SubscriptionDropReason.PersistentSubscriptionDeleted:
+                    Log.LogError(exception, "Persistent subscription was deleted. Stopping event publisher");
+                    _stop();
+                    break;
+                case SubscriptionDropReason.ConnectionClosed:
+                    Log.LogCritical(exception, "Underlying EventStore connection was closed. Stopping event publisher");
+                    _stop();
+                    break;
+                case SubscriptionDropReason.ProcessingQueueOverflow:
+                case SubscriptionDropReason.MaxSubscribersReached:
+                    Log.LogWarning(exception, "Transient error in EventStore connection. Attempting to resubscribe");
+                    TryToResubscribe().Wait();
```

Comment:
Wondering if there is a way where we don't have to block here...

Reply:
I think you'd need EventStore to change their API